File: p3391_submit.cpp
// P3391 【模板】文艺平衡树 —— 最优解（基于 reverse_bench 实测）
//
// 结构：sqrt_bitset（懒标记位图化块状链表）
//   - 块大小 B ≈ 2√n，>2B 拆块、<B/2 与邻居合并，块数保持 O(n/B)
//   - 懒反转标记打包进 u32 块描述符最高位：
//       整块区间翻转 = desc 区间反转（块序 + 标记随块走）+ vpxor 翻转标记
//       端部散块先物化再按元素反转
//   - 块本体放 arena + free-list 回收，无动态分配
// 实测（i9-13950HX）：n=m=1e5 随机区间约 20-40ms，比 splay/treap 快数倍。
// 可选 AVX2：编译时加 -mavx2 且 CPU 支持时启用 vpermd/vpxor，否则纯标量。
//
// 编译：g++ -O2 -std=c++17 p3391_submit.cpp -o p3391
//       （想开 SIMD：g++ -O2 -mavx2 -std=c++17 p3391_submit.cpp -o p3391）

#include <algorithm>
#include <cstdint>
#include <cstdio>
#include <cmath>
#include <vector>

#if defined(__AVX2__)
#include <immintrin.h>
#endif

using u32 = std::uint32_t;

static inline void reverse_scalar(u32* a, std::size_t l, std::size_t r) {
    while (l < r) {
        --r;
        u32 t = a[l];
        a[l] = a[r];
        a[r] = t;
        ++l;
    }
}

#if defined(__AVX2__)
static inline void reverse_avx2(u32* a, std::size_t l, std::size_t r) {
    const __m256i idx = _mm256_set_epi32(0, 1, 2, 3, 4, 5, 6, 7);
    while (l + 8 <= r) {
        __m256i L = _mm256_loadu_si256((const __m256i*)(a + l));
        __m256i R = _mm256_loadu_si256((const __m256i*)(a + r - 8));
        _mm256_storeu_si256((__m256i*)(a + l), _mm256_permutevar8x32_epi32(R, idx));
        _mm256_storeu_si256((__m256i*)(a + r - 8), _mm256_permutevar8x32_epi32(L, idx));
        l += 8;
        r -= 8;
    }
    reverse_scalar(a, l, r);
}
#endif

// 区间懒标记翻转（desc 的 REV 位）
static inline void toggle_mask(u32* a, u32 l, u32 r, u32 mask) {
#if defined(__AVX2__)
    if (__builtin_cpu_supports("avx2")) {
        const __m256i m = _mm256_set1_epi32((int)mask);
        u32 i = l;
        for (; i + 8 <= r; i += 8) {
            __m256i x = _mm256_loadu_si256((const __m256i*)(a + i));
            x = _mm256_xor_si256(x, m);
            _mm256_storeu_si256((__m256i*)(a + i), x);
        }
        for (; i < r; ++i) a[i] ^= mask;
        return;
    }
#endif
    for (u32 i = l; i < r; ++i) a[i] ^= mask;
}

static inline void reverse_best(u32* a, std::size_t l, std::size_t r) {
#if defined(__AVX2__)
    if (__builtin_cpu_supports("avx2")) {
        reverse_avx2(a, l, r);
        return;
    }
#endif
    reverse_scalar(a, l, r);
}

struct SqrtBitset {
    static constexpr u32 REV = 0x80000000u;

    struct Block {
        std::vector<u32> v;
    };

    std::vector<Block> pool;    // arena：块本体，id = 下标
    std::vector<u32> free_ids;  // 合并后回收的槽位
    std::vector<u32> desc;      // 逻辑块顺序：pool_id | (懒反转 << 31)
    std::vector<u32> szs;       // 与 desc 平行的块大小
    u32 B;

    explicit SqrtBitset(u32 n) {
        B = std::max(64u, std::min(1024u, (u32)(std::sqrt((double)n) * 2.0)));
        for (u32 i = 0; i < n;) {
            u32 e = std::min(n, i + B);
            Block bl;
            bl.v.reserve(e - i);
            for (u32 x = i; x < e; ++x) bl.v.push_back(x + 1);
            pool.push_back(std::move(bl));
            desc.push_back((u32)pool.size() - 1);
            szs.push_back(e - i);
            i = e;
        }
    }

    std::pair<u32, u32> find(u32 k) const {
        for (u32 i = 0; i < (u32)desc.size(); ++i) {
            u32 s = szs[i];
            if (k < s) return {i, k};
            k -= s;
        }
        return {0, 0};
    }

    inline void materialize(u32 pos) {
        if (desc[pos] & REV) {
            u32 id = desc[pos] & ~REV;
            reverse_best(pool[id].v.data(), 0, szs[pos]);
            desc[pos] &= ~REV;
        }
    }

    void split_block(u32 pos) {
        u32 s = szs[pos];
        if (s <= 2 * B) return;
        u32 half = s / 2;
        u32 id = desc[pos] & ~REV;  // 端块必已物化
        Block nb;
        nb.v.assign(pool[id].v.begin() + half, pool[id].v.end());
        pool[id].v.resize(half);
        szs[pos] = half;
        u32 nid;
        if (!free_ids.empty()) {
            nid = free_ids.back();
            free_ids.pop_back();
            pool[nid].v = std::move(nb.v);
        } else {
            pool.push_back(std::move(nb));
            nid = (u32)pool.size() - 1;
        }
        desc.insert(desc.begin() + pos + 1, nid);
        szs.insert(szs.begin() + pos + 1, s - half);
    }

    void merge_small(u32 pos) {
        u32 s = szs[pos];
        if (s >= B / 2) return;
        if (pos > 0 && szs[pos - 1] + s <= 2 * B) {
            materialize(pos - 1);
            materialize(pos);
            u32 idL = desc[pos - 1] & ~REV;
            u32 idR = desc[pos] & ~REV;
            pool[idL].v.insert(pool[idL].v.end(), pool[idR].v.begin(), pool[idR].v.end());
            szs[pos - 1] += s;
            std::vector<u32>().swap(pool[idR].v);
            free_ids.push_back(idR);
            desc.erase(desc.begin() + pos);
            szs.erase(szs.begin() + pos);
            return;
        }
        if (pos + 1 < desc.size() && s + szs[pos + 1] <= 2 * B) {
            materialize(pos);
            materialize(pos + 1);
            u32 idL = desc[pos] & ~REV;
            u32 idR = desc[pos + 1] & ~REV;
            pool[idL].v.insert(pool[idL].v.end(), pool[idR].v.begin(), pool[idR].v.end());
            szs[pos] += szs[pos + 1];
            std::vector<u32>().swap(pool[idR].v);
            free_ids.push_back(idR);
            desc.erase(desc.begin() + pos + 1);
            szs.erase(szs.begin() + pos + 1);
        }
    }

    void rebalance(u32 pos) {
        if (szs[pos] > 2 * B)
            split_block(pos);
        else
            merge_small(pos);
    }

    void reverse(u32 l, u32 r) {  // 元素下标 [l, r]（含端点）
        auto [bi, oi] = find(l);
        auto [bj, oj] = find(r);
        if (bi == bj) {
            materialize(bi);
            u32 id = desc[bi] & ~REV;
            reverse_best(pool[id].v.data(), oi, (std::size_t)oj + 1);
            return;
        }
        materialize(bi);
        materialize(bj);
        u32 idA = desc[bi] & ~REV;
        u32 idD = desc[bj] & ~REV;
        Block& A = pool[idA];
        Block& D = pool[idD];
        u32 nr = oj + 1;  // 右散块参与翻转的前缀长度
        std::vector<u32> rp(D.v.begin(), D.v.begin() + nr);
        reverse_best(rp.data(), 0, rp.size());
        std::vector<u32> lp(A.v.begin() + oi, A.v.end());
        reverse_best(lp.data(), 0, lp.size());
        std::vector<u32> tail(D.v.begin() + nr, D.v.end());
        A.v.resize(oi);
        A.v.insert(A.v.end(), rp.begin(), rp.end());
        D.v = std::move(lp);
        D.v.insert(D.v.end(), tail.begin(), tail.end());
        szs[bi] = (u32)A.v.size();
        szs[bj] = (u32)D.v.size();
        // 中间整块：块序反转（vpermd，标记随块走）+ 区间懒标记翻转（vpxor）
        reverse_best(desc.data(), bi + 1, bj);
        reverse_best(szs.data(), bi + 1, bj);
        toggle_mask(desc.data(), bi + 1, bj, REV);
        rebalance(bj);
        rebalance(bi);
    }

    void collect(std::vector<u32>& out) const {
        out.clear();
        for (u32 i = 0; i < (u32)desc.size(); ++i) {
            const std::vector<u32>& v = pool[desc[i] & ~REV].v;
            if (desc[i] & REV)
                out.insert(out.end(), v.rbegin(), v.rend());
            else
                out.insert(out.end(), v.begin(), v.end());
        }
    }
};

int main() {
    int n, m;
    if (scanf("%d%d", &n, &m) != 2) return 0;
    SqrtBitset sol((u32)n);
    for (int i = 0; i < m; ++i) {
        int l, r;
        scanf("%d%d", &l, &r);
        sol.reverse((u32)l - 1, (u32)r - 1);
    }
    std::vector<u32> out;
    sol.collect(out);
    for (u32 i = 0; i < (u32)out.size(); ++i)
        printf("%u%c", out[i], i + 1 == out.size() ? '\n' : ' ');
    return 0;
}

File: p3391_submit.rs
// P3391 【模板】文艺平衡树 —— Rust 提交版（sqrt_bitset，与 C++ 版同构）
//
// 结构：懒标记位图化块状链表
//   - 块大小 B ≈ 2√n，>2B 拆块、<B/2 与邻居合并，块数保持 O(n/B)
//   - 懒反转标记打包进 u32 块描述符最高位：
//       整块区间翻转 = desc 区间反转（块序 + 标记随块走）+ 标记位翻转
//       端部散块先物化再按元素反转
//   - 块本体放 arena + free-list 回收
// 快速 I/O：read_to_end 一次读入 + 手写解析；输出一次性 write_all。
//
// 编译：rustc -O p3391_submit.rs -o p3391

use std::io::{self, Read, Write};

const REV: u32 = 0x8000_0000;

struct SqrtBitset {
    pool: Vec<Vec<u32>>,  // arena：块本体，id = 下标
    free_ids: Vec<u32>,   // 合并后回收的槽位
    desc: Vec<u32>,       // 逻辑块顺序：pool_id | (懒反转 << 31)
    szs: Vec<u32>,        // 与 desc 平行的块大小
    b: u32,
}

impl SqrtBitset {
    fn new(n: u32) -> Self {
        let b = ((n as f64).sqrt() * 2.0) as u32;
        let b = b.clamp(64, 1024);
        let mut pool = Vec::new();
        let mut desc = Vec::new();
        let mut szs = Vec::new();
        let mut i = 0u32;
        while i < n {
            let e = (i + b).min(n);
            pool.push((i + 1..=e).collect());
            desc.push((pool.len() - 1) as u32);
            szs.push(e - i);
            i = e;
        }
        SqrtBitset {
            pool,
            free_ids: Vec::new(),
            desc,
            szs,
            b,
        }
    }

    fn find(&self, mut k: u32) -> (u32, u32) {
        for (i, &s) in self.szs.iter().enumerate() {
            if k < s {
                return (i as u32, k);
            }
            k -= s;
        }
        (0, 0)
    }

    fn materialize(&mut self, pos: u32) {
        if self.desc[pos as usize] & REV != 0 {
            let id = (self.desc[pos as usize] & !REV) as usize;
            self.pool[id].reverse();
            self.desc[pos as usize] &= !REV;
        }
    }

    fn split_block(&mut self, pos: u32) {
        let s = self.szs[pos as usize];
        if s <= 2 * self.b {
            return;
        }
        let half = s / 2;
        let id = (self.desc[pos as usize] & !REV) as usize; // 端块必已物化
        let nb = self.pool[id].split_off(half as usize);
        self.szs[pos as usize] = half;
        let nid = if let Some(f) = self.free_ids.pop() {
            self.pool[f as usize] = nb;
            f
        } else {
            self.pool.push(nb);
            (self.pool.len() - 1) as u32
        };
        self.desc.insert(pos as usize + 1, nid);
        self.szs.insert(pos as usize + 1, s - half);
    }

    fn merge_small(&mut self, pos: u32) {
        let s = self.szs[pos as usize];
        if s >= self.b / 2 {
            return;
        }
        let p = pos as usize;
        if p > 0 && self.szs[p - 1] + s <= 2 * self.b {
            self.materialize(pos - 1);
            self.materialize(pos);
            let idr = self.desc[p] & !REV;
            let mut rhs = std::mem::take(&mut self.pool[idr as usize]);
            let lid = (self.desc[p - 1] & !REV) as usize;
            self.pool[lid].append(&mut rhs);
            self.szs[p - 1] += s;
            self.free_ids.push(idr);
            self.desc.remove(p);
            self.szs.remove(p);
            return;
        }
        if p + 1 < self.desc.len() && s + self.szs[p + 1] <= 2 * self.b {
            self.materialize(pos);
            self.materialize(pos + 1);
            let idr = self.desc[p + 1] & !REV;
            let mut rhs = std::mem::take(&mut self.pool[idr as usize]);
            let lid = (self.desc[p] & !REV) as usize;
            self.pool[lid].append(&mut rhs);
            self.szs[p] += self.szs[p + 1];
            self.free_ids.push(idr);
            self.desc.remove(p + 1);
            self.szs.remove(p + 1);
        }
    }

    fn rebalance(&mut self, pos: u32) {
        if self.szs[pos as usize] > 2 * self.b {
            self.split_block(pos);
        } else {
            self.merge_small(pos);
        }
    }

    fn reverse(&mut self, l: u32, r: u32) {
        let (bi, oi) = self.find(l);
        let (bj, oj) = self.find(r);
        if bi == bj {
            self.materialize(bi);
            let id = (self.desc[bi as usize] & !REV) as usize;
            self.pool[id][oi as usize..=oj as usize].reverse();
            return;
        }
        self.materialize(bi);
        self.materialize(bj);
        let ai = (self.desc[bi as usize] & !REV) as usize;
        let di = (self.desc[bj as usize] & !REV) as usize;
        let nr = (oj + 1) as usize;
        let mut rp = self.pool[di][..nr].to_vec();
        rp.reverse();
        let mut lp = self.pool[ai][oi as usize..].to_vec();
        lp.reverse();
        let tail = self.pool[di][nr..].to_vec();
        self.pool[ai].truncate(oi as usize);
        self.pool[ai].extend_from_slice(&rp);
        self.pool[di] = lp;
        self.pool[di].extend_from_slice(&tail);
        self.szs[bi as usize] = self.pool[ai].len() as u32;
        self.szs[bj as usize] = self.pool[di].len() as u32;
        // 中间整块：块序反转（标记随块走）+ 区间懒标记翻转
        self.desc[(bi + 1) as usize..bj as usize].reverse();
        self.szs[(bi + 1) as usize..bj as usize].reverse();
        for d in &mut self.desc[(bi + 1) as usize..bj as usize] {
            *d ^= REV;
        }
        self.rebalance(bj);
        self.rebalance(bi);
    }

    fn collect(&self, out: &mut Vec<u32>) {
        out.clear();
        for i in 0..self.desc.len() {
            let d = self.desc[i];
            let v = &self.pool[(d & !REV) as usize];
            if d & REV != 0 {
                out.extend(v.iter().rev());
            } else {
                out.extend_from_slice(v);
            }
        }
    }
}

// ---------------- 快速 I/O ----------------
struct FastIn {
    buf: Vec<u8>,
    pos: usize,
}

impl FastIn {
    fn new() -> Self {
        let mut buf = Vec::new();
        io::stdin().lock().read_to_end(&mut buf).expect("read stdin");
        FastIn { buf, pos: 0 }
    }

    fn next_u32(&mut self) -> u32 {
        let b = &self.buf;
        let mut p = self.pos;
        while p < b.len() && b[p] <= b' ' {
            p += 1;
        }
        let mut x = 0u32;
        while p < b.len() && b[p] > b' ' {
            x = x * 10 + (b[p] - b'0') as u32;
            p += 1;
        }
        self.pos = p;
        x
    }
}

fn append_u32(out: &mut Vec<u8>, mut x: u32) {
    let mut tmp = [0u8; 10];
    let mut i = tmp.len();
    loop {
        i -= 1;
        tmp[i] = b'0' + (x % 10) as u8;
        x /= 10;
        if x == 0 {
            break;
        }
    }
    out.extend_from_slice(&tmp[i..]);
}

fn main() {
    let mut inp = FastIn::new();
    let n = inp.next_u32();
    let m = inp.next_u32();
    let mut sol = SqrtBitset::new(n);
    for _ in 0..m {
        let l = inp.next_u32();
        let r = inp.next_u32();
        sol.reverse(l - 1, r - 1);
    }
    let mut vals = Vec::with_capacity(n as usize);
    sol.collect(&mut vals);
    let mut out = Vec::with_capacity(n as usize * 8);
    for (i, &v) in vals.iter().enumerate() {
        if i > 0 {
            out.push(b' ');
        }
        append_u32(&mut out, v);
    }
    out.push(b'\n');
    io::stdout().lock().write_all(&out).expect("write stdout");
}
